### ISSUE TITLE
Inconsistent model inference

### ISSUE THREAD

**Issue by sipan17** [NVIDIA/NeMo#3229]
**Describe the bug**

Outputs different predictions for the same input depending batch. I suspect the problem is related to pad tokens not being masked properly.

**Steps/Code to reproduce bug**

```
import nemo.collections.asr as nemo_asr

model = nemo_asr.models.ASRModel.from_pretrained(model_name="stt_en_conformer_ctc_large")
model.transcribe(audio_paths[:2], batch_size=1)

# output
# ['take care of over excitement and endeavor to keep a quiet mind even for your health it is the best advice that can be given to you # # your moral and spiritual improvement will then keep pace with the culture of your intellectual powers',
# 'ab crowd']

model.transcribe(audio_paths[:2], batch_size=2)
# output
# ['take care of over excitement and endeavor to keep a quiet mind even for your health it is the best advice that can be given to you # # your moral and spiritual improvement will then keep pace with the culture of your intellectual powers',
# 'crow']

# the provided example occurred with audio files from the libritts100 dataset namely (3575_170457_000032_000001.wav and 6829_68771_000042_000001.wav), but I am quite sure it can be reproduced with other audio files as well. 
```

**Expected behavior**

Outputs should be identical independent of batch size or items in batch.

**Environment overview**

Linux Ubuntu 20.04
Python version: 3.8
PyTorch version: 1.10.0+cu113
Nemo version: 1.4.0
CUDA/cuDNN version: 11
GPU model and memory: RTX 3070, Tesla V100-SXM2-16GB

**Comment by VahidooX** [comment]
We mask the paddings to a good extent. We even mask them before the convolutions with kernel size of larger than one but paddings still can affect the results. Some operations like batch normalization do not ignore the paddings which may result in slightly different outputs. This issue is common among most ASR models. This example is a little extreme as a large sample is in the same batch with a very small one. We have evaluated our models with both batch size =1 and larger batches and usually accuracy difference on test sets are less than 0.1%. Is the accuracy you get significantly different on the whole set?

Which one is the correct transcription?

**Comment by sipan17** [comment]
I know it's hard to use masks with standard layers like batch norm, but in this case, maybe it's better to implement them by hand to correctly apply masks and do not let padding affect the output. Or as a hack, you can sort the list by durations internally, not to have such big differences, and after transcribing all files restore the original order.

Both transcriptions are wrong. The ground truth transcription is "I abhor crowds"

**Comment by titu1994** [comment]
We won't be using custom norm layers since that degrades compatibility with multi node training + mixed precision support. As mentioned, even LayerNorm is not a cureall, and this issue persists in all ASR models irrespective of framework. 

Sorting files by duration would mean precomputing the duration of all audio files, a procedure which has negligible benefit at exorbitant time cost.

**Comment by sipan17** [comment]
I measured on 43k audios of libritts100 dataset, it took 3 minutes with 8 processes, which might be preferred to error coming from padding. Anyway, it's just a solution that came to my mind.

**Comment by titu1994** [comment]
It would take two pases over the audio to measure just the durations. This doesn't scale for large datasets or multi node inference (parallel transcribe script). 

We use batch size 1 for academic settings - when publishing numbers for a paper cause it's a one off anyway. For all other settings we use batch size 32 or more (for RNNTs as high as 256) and accept a 0.1% loss of WER.
